containers/src/Data/Set/Internal.hs

```diff
@@ -1855,8 +1851,25 @@ cartesianProduct :: Set a -> Set b -> Set (a, b)
 -- Unfortunately, this is much slower in practice, at least when the sets are
 -- constructed from ascending lists. I tried doing the same thing using a
 -- known-length (perfect balancing) variant of fromDistinctAscList, but it
--- still didn't come close to the performance of Kmett's version in my very
--- informal tests.
+-- still didn't come close to the performance of the implementation we use in my
+-- very informal tests.
+--
+-- The implementation we use (slightly modified from one that Edward Kmett
+-- hacked together) is also optimal but performs better in practice. We map
+-- each element a in as to a set made up of (a,b) for every element b in bs,
+-- taking O(nm) overall. Then we merge these sets up the tree of as, which takes
+-- O(n log m). A brief sketch of proof for the latter:
+--
+-- There are O(n/2^i) nodes at level i of the tree. At a node at level i, we
+-- merge the merged left set, current set, and merged right set into a set of
+-- size O(2^i*m). This is done in a logarithmic time of O(i + log m). Over all
```

Comment:
Where does `i + log m` come from?

Reply:
$\log_2(2^i \cdot m) = i  + \log_2 m$

Reply:
Could you please specify exactly what you mean by "level $i$"? That seems to refer to the distance from a leaf node in some fashion, but readers shouldn't have to calculate things to figure that out.

Reply:
Ah yes, done. I think usually "level" goes the opposite direction, so we really should define it here.

Reply:
I don't like the phrasing "a logarithmic time". How about something like "This takes time O(log (2^i*m)) = O(i + log m)"?

Reply:
Done.